indexing-service/src/main/java/io/druid/indexing/overlord/HeapMemoryTaskStorage.java

```diff
@@ -198,6 +197,30 @@ public void addLock(final String taskid, final TaskLock taskLock)
   }
 
   @Override
+  public void upgradeLock(String taskid, TaskLock taskLock)
+  {
+    giant.lock();
+
+    try {
+      Preconditions.checkNotNull(taskid, "taskId");
+      Preconditions.checkNotNull(taskLock, "taskLock");
+      TaskLock taskLockToRemove = taskLock.withPreemptive(!taskLock.isPreemptive());
```

Comment:
if the taskLock is already non-preemptible, I think we should just check that it's in the taskLocks, and if so, return. This way the method is idempotent, which would be good, given that some of the existing tasks upgrade the same lock multiple times.